pkg/scheduler/reconciliation/reconciliation_test.go

```diff
@@ -729,3 +729,93 @@ func dbConnection(t *testing.T) db.Connection {
 	}
 	return dbConn
 }
+
+func TestReconciliationParallel(t *testing.T) {
+
+	t.Run("Create multiple instances of single reconciliations in parallel", func(t *testing.T) {
+		repo := newPersistentRepository(t)
+
+		inventory, err := cluster.NewInventory(dbConnection(t), true, cluster.MetricsCollectorMock{})
+		require.NoError(t, err)
+
+		errChannel := make(chan error, 100)
+		mockClusterState, _ := createClusterStates(t, inventory)
+
+		defer func() {
+			require.NoError(t, inventory.Delete(mockClusterState.Cluster.RuntimeID))
+		}()
+
+		startAt := time.Now().Add(1 * time.Second)
+		for i := 0; i < 50; i++ {
+			go func() {
+				time.Sleep(time.Until(startAt))
+				_, err := repo.CreateReconciliation(mockClusterState, nil)
+				if err != nil {
+					errChannel <- err
+				}
+			}()
+		}
+		time.Sleep(5 * time.Second)
+		require.Equal(t, 49, len(errChannel))
+	})
+
+	t.Run("Update single reconciliation in multiple parallel threads", func(t *testing.T) {
+		repo := newPersistentRepository(t)
+
+		inventory, err := cluster.NewInventory(dbConnection(t), true, cluster.MetricsCollectorMock{})
+		require.NoError(t, err)
+
+		errChannel := make(chan error, 100)
+		mockClusterState, _ := createClusterStates(t, inventory)
+
+		defer func() {
+			require.NoError(t, inventory.Delete(mockClusterState.Cluster.RuntimeID))
+		}()
+
+		recon, err := repo.CreateReconciliation(mockClusterState, nil)
+		require.NoError(t, err)
+		allOperations, err := repo.GetOperations(recon.SchedulingID)
+
+		startAt := time.Now().Add(1 * time.Second)
+		for i := 0; i < 50; i++ {
+			go func() {
+				time.Sleep(time.Until(startAt))
+				err = repo.UpdateOperationState(recon.SchedulingID, allOperations[0].CorrelationID, model.OperationStateError, "")
+				if err != nil {
+					errChannel <- err
+				}
+			}()
+		}
+		time.Sleep(5 * time.Second)
+		require.Equal(t, 49, len(errChannel))
+	})
+
+	t.Run("Mark single reconciliation as finished in multiple parallel threads", func(t *testing.T) {
+		repo := newPersistentRepository(t)
+
+		inventory, err := cluster.NewInventory(dbConnection(t), true, cluster.MetricsCollectorMock{})
+		require.NoError(t, err)
+
+		errChannel := make(chan error, 100)
+		mockClusterState, _ := createClusterStates(t, inventory)
+
+		defer func() {
+			require.NoError(t, inventory.Delete(mockClusterState.Cluster.RuntimeID))
+		}()
+
+		recon, err := repo.CreateReconciliation(mockClusterState, nil)
+
+		startAt := time.Now().Add(1 * time.Second)
+		for i := 0; i < 50; i++ {
+			go func() {
+				time.Sleep(time.Until(startAt))
+				err = repo.FinishReconciliation(recon.SchedulingID, mockClusterState.Status)
+				if err != nil {
+					errChannel <- err
+				}
+			}()
+		}
+		time.Sleep(5 * time.Second)
+		require.Equal(t, 49, len(errChannel))
```

Comment:
Please also quickly verify that the reconciliation was updated to be finished.

We forgot one thing to verify: the the public functions of a transition (see [transition.go](https://github.com/kyma-incubator/reconciler/blob/main/pkg/scheduler/service/transition.go))should also be checked to handle race-conditions properly (sorry, forgot that part)